phoenix-core/src/it/java/org/apache/phoenix/end2end/OnDuplicateKey2IT.java

```diff
@@ -118,6 +139,177 @@ public void testIgnoreReturnValue() throws Exception {
         conn.close();
     }
 
+    @Test
+    public void testReturnRowResult() throws Exception {
+        Assume.assumeTrue("Set correct result to RegionActionResult on hbase versions " +
+                "2.4.18+, 2.5.9+, and 2.6.0+", isSetCorrectResultEnabledOnHBase());
+
+        Properties props = PropertiesUtil.deepCopy(TEST_PROPERTIES);
+        String sample1 = getJsonString("json/sample_01.json");
+        String sample2 = getJsonString("json/sample_02.json");
+        BsonDocument bsonDocument1 = RawBsonDocument.parse(sample1);
+        BsonDocument bsonDocument2 = RawBsonDocument.parse(sample2);
+        try (Connection conn = DriverManager.getConnection(getUrl(), props)) {
+            conn.setAutoCommit(true);
+            String tableName = generateUniqueName();
+            String ddl = "CREATE TABLE " + tableName
+                    + "(PK1 VARCHAR, PK2 DOUBLE NOT NULL, PK3 VARCHAR, COUNTER1 DOUBLE,"
+                    + " COUNTER2 VARCHAR,"
+                    + " COL3 BSON, COL4 INTEGER, CONSTRAINT pk PRIMARY KEY(PK1, PK2, PK3))";
+            conn.createStatement().execute(ddl);
+            createIndex(conn, tableName);
+
+            String upsertSql = "UPSERT INTO " + tableName + " (PK1, PK2, PK3, COUNTER1, COL3, COL4)"
+                    + " VALUES('pk000', -123.98, 'pk003', 1011.202, ?, 123) ON DUPLICATE KEY " +
+                    "IGNORE";
+            validateReturnedRowAfterUpsert(conn, upsertSql, tableName, 1011.202, null, true,
+                    bsonDocument1, bsonDocument1, 123);
+
+            upsertSql =
+                    "UPSERT INTO " + tableName + " (PK1, PK2, PK3, COUNTER1) "
+                            + "VALUES('pk000', -123.98, 'pk003', 0) ON DUPLICATE"
+                            + " KEY IGNORE";
+            validateReturnedRowAfterUpsert(conn, upsertSql, tableName, 1011.202, null, false,
+                    null, bsonDocument1, 123);
+
+            upsertSql =
+                    "UPSERT INTO " + tableName
+                            + " (PK1, PK2, PK3, COUNTER1, COUNTER2) VALUES('pk000', -123.98, "
+                            + "'pk003', 234, 'col2_000')";
+            validateReturnedRowAfterUpsert(conn, upsertSql, tableName, 234d, "col2_000", true,
```

Comment:
@tkhurana this is example of partial row update where the API returns full row state.

Reply:
COL3 and COL4 becomes null because the UPSERT updated only COUNTER1 and COUNTER2

Reply:
Why will COL3 and COL4 become null if they are not being updated ?

Reply:
Let me fix this, good catch, i did not realize that for non-atomic (without using ON DUPLICATE clause) this was an oversight.